From fd8bcefa890ddafc142d0a80e7aee4938a7c5fe9 Mon Sep 17 00:00:00 2001
From: Nariman Poushin <nariman@opensource.wolfsonmicro.com>
Date: Wed, 23 Jul 2014 14:55:55 +0100
Subject: [PATCH 265/351] switch: arizona: Add support for external hpdet
 resistor

Some platforms have fixed resistors in series with HPDET to
suppress pops. Make sure we can take this in to account
when determining playback path impedance.

Change-Id: I5fb80b7a430f17981997adac358d7138b3fd0b6a
Signed-off-by: Nariman Poushin <nariman@opensource.wolfsonmicro.com>
---
 drivers/mfd/arizona-core.c        |    3 +++
 drivers/switch/switch-arizona.c   |   17 +++++++++++++++++
 include/linux/mfd/arizona/pdata.h |    2 ++
 3 files changed, 22 insertions(+)

diff --git a/drivers/mfd/arizona-core.c b/drivers/mfd/arizona-core.c
index ce1be6e0..3027384 100644
--- a/drivers/mfd/arizona-core.c
+++ b/drivers/mfd/arizona-core.c
@@ -875,6 +875,9 @@ static int arizona_of_get_core_pdata(struct arizona *arizona)
 
 	arizona_of_read_u32(arizona, "wlf,wm5102t-output-pwr", false,
 				&pdata->wm5102t_output_pwr);
+
+	arizona_of_read_u32(arizona, "wlf,hpdet-ext-res", false,
+				&pdata->hpdet_ext_res);
 	return 0;
 }
 
diff --git a/drivers/switch/switch-arizona.c b/drivers/switch/switch-arizona.c
index 2bc387f..7f0ff24 100644
--- a/drivers/switch/switch-arizona.c
+++ b/drivers/switch/switch-arizona.c
@@ -578,8 +578,25 @@ static int arizona_hpdet_read(struct arizona_extcon_info *info)
 		}
 	}
 
+	if (info->arizona->pdata.hpdet_ext_res) {
+
+		if (info->arizona->pdata.hpdet_ext_res >=  val) {
+			dev_err(arizona->dev,
+				"External resistor (%d) >= measurement (%d)\n",
+				info->arizona->pdata.hpdet_ext_res,
+				val);
+		} else {
+			dev_dbg(arizona->dev,
+				"Compensating for external %d ohm resistor\n",
+				info->arizona->pdata.hpdet_ext_res);
+
+			val -= info->arizona->pdata.hpdet_ext_res;
+		}
+	}
+
 	dev_dbg(arizona->dev, "HP impedance %d ohms\n", val);
 
+
 	return val;
 }
 
diff --git a/include/linux/mfd/arizona/pdata.h b/include/linux/mfd/arizona/pdata.h
index 35978f7..61cc7ac 100644
--- a/include/linux/mfd/arizona/pdata.h
+++ b/include/linux/mfd/arizona/pdata.h
@@ -252,6 +252,8 @@ struct arizona_pdata {
 	struct wm_adsp_fw_defs *fw_defs[ARIZONA_MAX_DSP];
 	int num_fw_defs[ARIZONA_MAX_DSP];
 
+	/** Some platforms add a series resistor for hpdet to suppress pops */
+	int hpdet_ext_res;
 };
 
 #endif
-- 
1.7.9.5

